File: c/problems/problem#368.c
#include <stdlib.h>

int comp(const void *elem1, const void *elem2) {
  int f = *((int *)elem1);
  int s = *((int *)elem2);
  if (f > s)
    return 1;
  if (f < s)
    return -1;
  return 0;
}

int *largestDivisibleSubset(int *nums, int n, int *returnSize) {
  qsort(nums, n, sizeof(int), comp);

  int *next = malloc(n * sizeof(int));
  int *len = malloc(n * sizeof(int));
  for (int i = 0; i < n; i++) {
    len[i] = 1;
    next[i] = -1;
  }
  for (int i = n - 2; i >= 0; i--) {
    for (int j = i + 1; j < n; j++) {
      if (nums[j] % nums[i] == 0 && len[i] < len[j] + 1) {
        len[i] = len[j] + 1;
        next[i] = j;
      }
    }
  }
  int max_len = 0;
  int x = 0;
  for (int i = 0; i < n; i++) {
    if (len[i] > max_len) {
      max_len = len[i];
      x = i;
    }
  }
  int *ans = malloc(max_len * sizeof(int));
  *returnSize = max_len;
  int i = 0;

  while (x != -1) {
    ans[i] = nums[x];
    x = next[x];
    i++;
  }

  free(next);
  free(len);

  return ans;
}

File: go/problem#368.go
package main

import "slices"

func largestDivisibleSubset(nums []int) []int {
	n := len(nums)
	slices.Sort(nums)
	ans := []int{}
	next := make([]int, n)
	s_len := make([]int, n)
	s_len[n-1] = 1
	next[n-1] = -1
	for i := n - 2; i >= 0; i-- {
		s_len[i] = 1
		next[i] = -1
		for j := i + 1; j < n; j++ {
			if nums[j]%nums[i] == 0 && s_len[j]+1 > s_len[i] {
				s_len[i] = s_len[j] + 1
				next[i] = j
			}
		}
	}
	max_len := 0
	for i := range n {
		if max_len < s_len[i] {
			max_len = s_len[i]
		}
	}
	x := 0
	for i := range n {
		if s_len[i] == max_len {
			x = i
			break
		}
	}
	for x != -1 {
		ans = append(ans, nums[x])
		x = next[x]
	}
	return ans
}
